### PR TITLE
Improve typing for plugins

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #10163
Requested by #10682 
Part of #7037 

cc @borda @justusschock @kaushikb11 @awaelchli @akihironitta @ananthsub @daniellepintz 

## What does this PR do?

Fixes

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

### CODE REVIEW

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 if _TORCH_GREATER_EQUAL_1_10:
     from torch import autocast
 else:
-    from torch.cuda.amp import autocast
+    from torch.cuda.amp import autocast  # type: ignore[misc]
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/native_amp.py:30]: @carmocca here is an example of where mypy would not be aware of different torch versions

Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py]: Fixed

Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py]: You are too good at this ...

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 2]
```diff
@@ -147,7 +147,7 @@ def optimizer_step(
         """Hook to run the optimizer step."""
         if isinstance(model, pl.LightningModule):
             closure = partial(self._wrap_closure, model, optimizer, optimizer_idx, closure)
-        optimizer.step(closure=closure, **kwargs)
+        optimizer.step(closure=closure, **kwargs)  # type: ignore[call-arg]
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/precision_plugin.py:150]: I've always found passing `**kwargs` here weird. I guess it's useful for weird Optimizer implementations which violate the Liskov Substitution Principle.

One way we could get around it is doing

```python
    @runtime_checkable
    class _Optimizer(Protocol):
        def step(self, closure: Callable[[], Any], **kwargs: Any) -> Any:
            ...

    def optimizer_step(
        self,
        model: Union["pl.LightningModule", Module],
        optimizer: _Optimizer,
        optimizer_idx: int,
        closure: Callable[[], Any],
        **kwargs: Any,
    ) -> None:
```

Or maybe I'm just too excited about having learned a new thing and want to use it where I shouldn't.

Do you have any thoughts about this @ananthsub?

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/precision_plugin.py:150]: I think it is fine for us to type: ignore because we know this is about custom Optimizers and their step method. I can add a comment if you want. 

If you insist, please go ahead and commit it directly since you know best

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/precision_plugin.py:150]: It was more of a side comment. This does not block this PR

**Review by akihironitta — APPROVED** [review]
